include/pybind11/pytypes.h

```diff
@@ -375,31 +388,101 @@ PYBIND11_NAMESPACE_END(detail)
 /// python).
 class PYBIND11_EXPORT_EXCEPTION error_already_set : public std::runtime_error {
 public:
-    /// Constructs a new exception from the current Python error indicator, if any.  The current
-    /// Python error indicator will be cleared.
-    error_already_set() : std::runtime_error(detail::error_string()) {
+    /// Fetches the current Python exception (using PyErr_Fetch()), which will clear the
+    /// current Python error indicator.
+    error_already_set() : std::runtime_error("") {
+        if (!PyErr_Occurred()) {
+            pybind11_fail("Internal error: pybind11::detail::error_already_set called while "
+                          "Python error indicator not set.");
+        }
         PyErr_Fetch(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
     }
 
     error_already_set(const error_already_set &) = default;
-    error_already_set(error_already_set &&) = default;
+
+    /// Moving the members one-by-one to be able to specify noexcept.
+    error_already_set(error_already_set &&e) noexcept
+        : std::runtime_error(std::move(e)), m_type{std::move(e.m_type)},
+          m_value{std::move(e.m_value)}, m_trace{std::move(e.m_trace)}, m_lazy_what{std::move(
+                                                                            e.m_lazy_what)} {};
 
     inline ~error_already_set() override;
 
-    /// Give the currently-held error back to Python, if any.  If there is currently a Python error
-    /// already set it is cleared first.  After this call, the current object no longer stores the
-    /// error variables (but the `.what()` string is still available).
+    /// NOTE: This function may have the side-effect of normalizing the held Python exception
+    ///       (if it is not normalized already).
+    const char *what() const noexcept override {
+        if (m_lazy_what.empty()) {
+            try {
+                m_lazy_what = detail::error_string(m_type.ptr(), m_value.ptr(), m_trace.ptr());
+                // Negate the if condition to test the catch(...) block below.
+                if (m_lazy_what.empty()) {
+                    throw std::runtime_error(
```

Comment:
Several issues with this: 
1. This should be a pybind11_fail, right?
2. This is inconsistent with the catch statement std::terminate. They should have the same failure pass, (either they should have the fancy writing to stdout and stderr or they should both just throw exceptions in the noexcept area).

Reply:
3. This is no reason for this to be in try block at all. It should just be after the catch statement then.

Reply:
@rwgk Yeah here, I mean the message text from the std::runtime error here is completely lost, right?